pkg/controller/shoot/shoot_control_reconcile.go

```diff
@@ -52,58 +52,226 @@ func (c *defaultControl) reconcileShoot(o *operation.Operation, operationType ga
-		waitUntilEtcdReady                      = f.AddTask(botanist.WaitUntilEtcdReady, 0, deployETCD)
-		deployCloudProviderConfig               = f.AddTask(hybridBotanist.DeployCloudProviderConfig, defaultRetry, deployInfrastructure)
-		deployKubeAPIServer                     = f.AddTask(hybridBotanist.DeployKubeAPIServer, defaultRetry, deploySecrets, deployETCD, waitUntilEtcdReady, waitUntilKubeAPIServerServiceIsReady, deployCloudProviderConfig)
-		_                                       = f.AddTask(hybridBotanist.DeployKubeControllerManager, defaultRetry, deploySecrets, deployCloudProviderConfig, deployKubeAPIServer)
-		_                                       = f.AddTask(hybridBotanist.DeployKubeScheduler, defaultRetry, deploySecrets, deployKubeAPIServer)
-		waitUntilKubeAPIServerIsReady           = f.AddTask(botanist.WaitUntilKubeAPIServerReady, 0, deployKubeAPIServer)
-		initializeShootClients                  = f.AddTask(botanist.InitializeShootClients, 2*time.Minute, waitUntilKubeAPIServerIsReady)
-		deployMachineControllerManager          = f.AddTaskConditional(botanist.DeployMachineControllerManager, defaultRetry, isCloud, initializeShootClients)
-		reconcileMachines                       = f.AddTaskConditional(hybridBotanist.ReconcileMachines, defaultRetry, isCloud, deployMachineControllerManager, deployInfrastructure, initializeShootClients)
-		deployKubeAddonManager                  = f.AddTask(hybridBotanist.DeployKubeAddonManager, defaultRetry, initializeShootClients, deployInfrastructure)
-		_                                       = f.AddTask(shootCloudBotanist.DeployKube2IAMResources, defaultRetry, deployInfrastructure)
-		_                                       = f.AddTaskConditional(botanist.EnsureIngressDNSRecord, 10*time.Minute, managedDNS, deployKubeAddonManager)
-		waitUntilVPNConnectionExists            = f.AddTaskConditional(botanist.WaitUntilVPNConnectionExists, 0, !o.Shoot.Hibernated, deployKubeAddonManager, reconcileMachines)
-		applyCreateHook                         = f.AddTask(seedCloudBotanist.ApplyCreateHook, defaultRetry, waitUntilVPNConnectionExists)
-		deploySeedMonitoring                    = f.AddTask(botanist.DeploySeedMonitoring, defaultRetry, waitUntilKubeAPIServerIsReady, initializeShootClients, waitUntilVPNConnectionExists, reconcileMachines, applyCreateHook)
-		_                                       = f.AddTask(botanist.DeployClusterAutoscaler, defaultRetry, reconcileMachines, deployKubeAddonManager, deploySeedMonitoring)
+		defaultTimeout  = 30 * time.Second
+		defaultInterval = 5 * time.Second
+		managedDNS      = o.Shoot.Info.Spec.DNS.Provider != gardenv1beta1.DNSUnmanaged
+		isCloud         = o.Shoot.Info.Spec.Cloud.Local == nil
+
+		g = flow.NewGraph("Shoot cluster reconciliation")
+
+		deployNamespace = g.Add(flow.Task{
+			Name: "Deploy Namespace",
+			Fn: flow.TaskFn(botanist.DeployNamespace).
+				RetryUntilTimeout(defaultInterval, defaultTimeout),
```

Comment:
Let's put that one line above to improve readability (same for the following tasks).